server/aesdsocket.c

```diff
@@ -2,260 +2,285 @@
-
-        syslog(LOG_INFO, "connection accepted:- %s", inet_ntoa(client_addr.sin_addr));
-
-        size_t total_received = 0;
-        ssize_t received;
-
-
-        memset(resources->buffer, 0, resources->buffer_size);
-
-
-        while ((received = recv(resources->client_fd , resources->buffer + total_received  , resources->buffer_size - total_received - 1  , 0) ) >  0) 
-        {
-            total_received += received;
-
-
-            if (total_received >= resources->buffer_size - 1) 
-            {
-            
-                size_t new_size = resources->buffer_size * 2;
-                char *new_buffer = realloc(resources->buffer, new_size);
-                if (!new_buffer) {
-                    syslog(LOG_ERR, "failed_realloc ");
-                    cleanup(SIGTERM);
-                }
-                resources->buffer = new_buffer;
-                resources->buffer_size = new_size;
-            }
-
-
-            if (memchr(resources->buffer, '\n', total_received)) {
-                break;
-            }
+        client_thread_t *ct = calloc(1, sizeof(client_thread_t));
+        if (!ct) {
+            syslog(LOG_ERR, "calloc() for client_thread_t failed");
+            close(client_fd);
+            continue;
         }
-
-        if (received == -1) {
-            syslog(LOG_ERR, "receive failed: %s", strerror(errno));
-        } else if (total_received > 0) {
-
-            if (fwrite(resources->buffer, 1, total_received, resources->data_file) != total_received) {
-                syslog(LOG_ERR, "write to file failed: %s", strerror(errno));
-            }
-            fflush(resources->data_file);
-
-
-            fseek(resources->data_file, 0, SEEK_SET);
-            while (fgets(resources->buffer, resources->buffer_size, resources->data_file)) {
-                send(resources->client_fd, resources->buffer, strlen(resources->buffer), 0);
-            }
-            fseek(resources->data_file, 0, SEEK_END);
+        ct->client_fd = client_fd;
+        ct->client_addr = clientaddr;
+        SLIST_INSERT_HEAD(&g_client_list, ct, entries);
+        if (pthread_create(&ct->thread_id, NULL, client_thread_func, ct) != 0) {
+            syslog(LOG_ERR, "pthread_create for client failed");
+            close(client_fd);
+            SLIST_REMOVE(&g_client_list, ct, client_thread_s, entries);
+            free(ct);
         }
-
-        syslog(LOG_INFO, "connection clsed:- %s", inet_ntoa(client_addr.sin_addr));
-        close(resources->client_fd);
-        resources->client_fd = -1;
     }
-
-    cleanup(SIGTERM);
+    syslog(LOG_INFO, "shutting down aesdsocket...........................................");
+    if (g_resources.listen_fd >= 0) {
+        close(g_resources.listen_fd);
+        g_resources.listen_fd = -1;
+    }
+    client_thread_t *it = NULL, *tmp = NULL;
+    for (it = SLIST_FIRST(&g_client_list); it != NULL; it = tmp) {
+        tmp = SLIST_NEXT(it, entries);
+        pthread_join(it->thread_id, NULL);
```

Comment:
@Sriramz2002 see discussion in class about pthread_join outside the while loop here and what this means for memory usage.